src/backend/distributed/operations/shard_cleaner.c

```diff
@@ -131,32 +131,62 @@ DropMarkedShards(bool waitForCleanupLock)
 			continue;
 		}
 
-		ShardPlacement *shardPlacement = LoadShardPlacement(placement->shardId,
-															placement->placementId);
-		ShardInterval *shardInterval = LoadShardInterval(shardPlacement->shardId);
-
-		ereport(LOG, (errmsg("dropping shard placement " INT64_FORMAT " of shard "
-							 INT64_FORMAT " on %s:%d after it was moved away",
-							 shardPlacement->placementId, shardPlacement->shardId,
-							 shardPlacement->nodeName, shardPlacement->nodePort)));
+		if (TryDropShard(placement))
+		{
+			removedShardCount++;
+		}
+	}
 
-		/* prepare sql query to execute to drop the shard */
-		StringInfo dropQuery = makeStringInfo();
-		char *qualifiedTableName = ConstructQualifiedShardName(shardInterval);
-		appendStringInfo(dropQuery, DROP_REGULAR_TABLE_COMMAND, qualifiedTableName);
+	return removedShardCount;
+}
 
-		List *dropCommandList = list_make2("SET LOCAL lock_timeout TO '1s'",
-										   dropQuery->data);
 
-		/* remove the shard from the node and the placement information */
+/*
+ * TryDropShard tries to drop the given shard placement and returns
+ * true on success. On failure, this method swallows errors and emits them
+ * as WARNINGs.
+ */
+static bool
+TryDropShard(GroupShardPlacement *placement)
+{
+	MemoryContext savedContext = CurrentMemoryContext;
+	ShardPlacement *shardPlacement = LoadShardPlacement(placement->shardId,
+														placement->placementId);
+	ShardInterval *shardInterval = LoadShardInterval(shardPlacement->shardId);
+
+	ereport(LOG, (errmsg("dropping shard placement " INT64_FORMAT " of shard "
+						 INT64_FORMAT " on %s:%d after it was moved away",
+						 shardPlacement->placementId, shardPlacement->shardId,
+						 shardPlacement->nodeName, shardPlacement->nodePort)));
+
+	/* prepare sql query to execute to drop the shard */
+	StringInfo dropQuery = makeStringInfo();
+	char *qualifiedTableName = ConstructQualifiedShardName(shardInterval);
+	appendStringInfo(dropQuery, DROP_REGULAR_TABLE_COMMAND, qualifiedTableName);
+
+	List *dropCommandList = list_make2("SET LOCAL lock_timeout TO '1ms'",
+									   dropQuery->data);
+	PG_TRY();
```

Comment:
Is there a way to do this without PG_TRY? Continuing after a PG_TRY has various issues, because some of the code might rely on the abort handler resetting global state before the next use, but this does not happen when we continue after PG_CATCH.

Reply:
Good point, With an optional command we can do that:https://github.com/citusdata/citus/pull/4919/files#diff-a4f78ce09ab799df4e30876818af2e55f2db8a11cdf932f98304e127961aecc8R519

I have tested this manually and it partially dropped some placements and failed on other ones (which were locked) as expected

Reply:
I can try to add a test to cover this